// graphql-svelte/dist/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.svqlConfig = void 0;
var GraphQL_1 = require("./src/GraphQL");
Object.defineProperty(exports, "GraphQL", { enumerable: true, get: function () { return GraphQL_1.GraphQL; } });
var reportCacheErrors_1 = require("./src/reportCacheErrors");
Object.defineProperty(exports, "reportCacheErrors", { enumerable: true, get: function () { return reportCacheErrors_1.reportCacheErrors; } });
var graphqlFetchOptions_1 = require("./src/graphqlFetchOptions");
Object.defineProperty(exports, "graphqlFetchOptions", { enumerable: true, get: function () { return graphqlFetchOptions_1.graphqlFetchOptions; } });
var hashObject_1 = require("./src/hashObject");
Object.defineProperty(exports, "hashObject", { enumerable: true, get: function () { return hashObject_1.hashObject; } });
var SubscribeQL_1 = require("./src/SubscribeQL");
Object.defineProperty(exports, "SubscribeQL", { enumerable: true, get: function () { return SubscribeQL_1.SubscribeQL; } });
var svqlConfig = __importStar(require("./src/svqlConfig"));
exports.svqlConfig = svqlConfig;
;
;
